src/v/cluster/cluster_discovery.cc

```diff
@@ -71,14 +71,74 @@ ss::future<node_id> cluster_discovery::determine_node_id() {
     co_return assigned_node_id;
 }
 
-vector<broker> cluster_discovery::initial_raft0_brokers() const {
+namespace {
+
+/**
+ * Search for the current node's advertised RPC address in the seed_servers.
+ * Precondition: emtpy_seed_starts_cluster=false
+ * \return Index of this node in seed_servers list if found, or empty if
+ * not.
+ * \throw runtime_error if seed_servers is empty
+ */
+std::optional<model::node_id::type> get_node_index_in_seed_servers() {
+    const std::vector<config::seed_server>& seed_servers
+      = config::node().seed_servers();
+    if (seed_servers.empty()) {
+        throw std::runtime_error(fmt_with_ctx(
+          fmt::format,
+          "Configuration error: seed_servers cannot be empty when "
+          "empty_seed_starts_cluster is false"));
+    }
+    const auto it = std::find_if(
+      seed_servers.cbegin(),
+      seed_servers.cend(),
+      [rpc_addr = config::node().advertised_rpc_api()](
+        const config::seed_server& seed_server) {
+          return rpc_addr == seed_server.addr;
+      });
+    if (it == seed_servers.cend()) {
+        return {};
+    }
+    return {boost::numeric_cast<model::node_id::type>(
+      std::distance(seed_servers.cbegin(), it))};
+}
+
+} // namespace
+
+ss::future<cluster_discovery::brokers>
+cluster_discovery::initial_seed_brokers() {
     // If configured as the root node, we'll want to start the cluster with
     // just this node as the initial seed.
-    if (is_cluster_founder()) {
-        // TODO: we should only return non-empty seed list if our log is empty.
-        return {make_self_broker(config::node())};
+    if (config::node().empty_seed_starts_cluster()) {
+        if (config::node().seed_servers().empty()) {
+            co_return brokers{make_self_broker(config::node())};
+        }
+        // Not a root
+        co_return brokers{};
     }
-    return {};
+    if (get_node_index_in_seed_servers()) {
```

Comment:
Should this actually be `is_cluster_founder()`? Then we'd even be able to avoid the duplicate call to `fetch_cluster_bootstrap_info()`.

Also it handles the case when another node has a cluster UUID (i.e. the local node is not a cluster founder).

Generally, it'd be great if instead of smearing various calls to figure out whether we're a founder around the bootstrap path (e.g. `stored_cluster_uuid`), we should figure up ASAP whether or not we should be founding a cluster. To that end, we make all the calls to storage and the bootstrap endpoints up front, and figure out whether a cluster UUID exists locally, or among the seed servers. From then on, we can keep that cached in `cluster_discovery` and have `cluster_discovery` handle behavioral changes based on whether we're a founder.